### PR TITLE
Human-readable revision describe

### PR DESCRIPTION
**Pull request by sixolet**
Revision describe based on new format for service describe. 

Non-verbose:
```
$ ./kn revision describe bar-bhndh-34 
Name:         bar-bhndh-34
Namespace:    default
Annotations:  autoscaling.knative.dev/target=-1
Age:          32d
Image:        gcr.io/knative-samples/helloworld-nodejs:latest (pinned to ae8c74)
Env:          ASDF=tame, BAZ=quux
Concurrency:  
  Limit:      1
  Target:     -1
CPU:          400m
Service:      bar

Conditions:  
  OK TYPE                  AGE REASON
  ++ Ready                 32d 
  ++ BuildSucceeded        32d 
  ++ ContainerHealthy      32d 
  ++ ResourcesAvailable    32d 
   I Active                32d NoTraffic
```

Verbose (note the extra info about service):
```
$ ./kn revision describe bar-bhndh-34 -v
Name:              bar-bhndh-34
Namespace:         default
Labels:            client.knative.dev/nonce=ygvkcsnvhj
                   serving.knative.dev/configuration=bar
                   serving.knative.dev/configurationGeneration=23
                   serving.knative.dev/service=bar
Annotations:       autoscaling.knative.dev/target=-1
                   client.knative.dev/user-image=gcr.io/knative-samples/helloworld-nodejs:latest
                   serving.knative.dev/lastPinned=1572895120
Age:               32d
Image:             gcr.io/knative-samples/helloworld-nodejs:latest (pinned to ae8c74)
Env:               ASDF=tame
                   BAZ=quux
Concurrency:       
  Limit:           1
  Target:          -1
CPU:               400m
Service:           bar
  Config Gen:      23
  Latest Created:  true
  Latest Ready:    true
  Traffic:         10%
  Tags:            latest

Conditions:  
  OK TYPE                  AGE REASON
  ++ Ready                 32d 
  ++ BuildSucceeded        32d 
  ++ ContainerHealthy      32d 
  ++ ResourcesAvailable    32d 
   I Active                32d NoTraffic (The target is not receiving traffic.)
```

### CODE REVIEW

File: pkg/kn/commands/revision/describe.go [hunk 1]
```diff
@@ -46,19 +62,225 @@ func NewRevisionDescribeCommand(p *commands.KnParams) *cobra.Command {
...
+	WritePort(dw, revision)
+	WriteEnv(dw, revision, printDetails)
+	WriteScale(dw, revision)
+	WriteConcurrencyOptions(dw, revision)
+	WriteResources(dw, revision)
+	serviceName, ok := revision.Labels[servingserving.ServiceLabelKey]
+	if ok {
+		serviceSection := dw.WriteAttribute("Service", serviceName)
+		if printDetails {
+			serviceSection.WriteAttribute("Config Gen", revision.Labels[servingserving.ConfigurationGenerationLabelKey])
```
Comment (rhuss) [inline pkg/kn/commands/revision/describe.go]: As we are in verbose mode anyway, could we expand this to "Configuration Generation" or maybe just "Generation" ? (compromise: "Config Generation")

File: pkg/kn/commands/describe.go [hunk 2]
```diff
@@ -180,6 +180,32 @@ func WriteConditions(dw printers.PrefixWriter, conditions []apis.Condition, prin
 	}
 }
 
+// Writer a slice compact (printDetails == false) in one line, or over multiple line
+// with key-value line-by-line (printDetails == true)
+func WriteSliceDesc(dw printers.PrefixWriter, s []string, label string, printDetails bool) {
```
Comment (rhuss) [inline pkg/kn/commands/describe.go]: Before we forget, should we then also remove https://github.com/knative/client/blob/aafdb9a2387b5a50d880a2ee563fcf2786217fc8/pkg/kn/commands/service/describe.go#L325-L345 and let `kn service describe` use this method ?

File: pkg/kn/commands/revision/describe.go [hunk 3]
```diff
@@ -46,19 +62,225 @@ func NewRevisionDescribeCommand(p *commands.KnParams) *cobra.Command {
...
+func shortenDigest(digest string) string {
+	match := imageDigestRegexp.FindStringSubmatch(digest)
+	if len(match) > 1 {
+		return string(match[1][:6])
+	}
+	return digest
+}
+
+// Format scale in the format "min ... max" with max = ∞ if not set
+func formatScale(minScale *int, maxScale *int) string {
```
Comment (rhuss) [inline pkg/kn/commands/revision/describe.go]: Should we move that up to a `FormatScale`, reusable by `service describe` ?

File: pkg/kn/commands/revision/describe.go [hunk 4]
```diff
@@ -46,19 +62,225 @@ func NewRevisionDescribeCommand(p *commands.KnParams) *cobra.Command {
...
+	}
+	return nil
+}
+
+func WriteConcurrencyOptions(dw printers.PrefixWriter, revision *v1alpha1.Revision) {
+	target := clientserving.ConcurrencyTarget(&revision.ObjectMeta)
+	limit := revision.Spec.ContainerConcurrency
+	if target != nil || limit != nil {
+		section := dw.WriteAttribute("Concurrency", "")
+		if limit != nil {
```
Comment (rhuss) [inline pkg/kn/commands/revision/describe.go]: I think we shouldn't write out the limit if its `== 0`, as it is the case by default.

Comment (rhuss) [reply pkg/kn/commands/revision/describe.go]: That's my output with setting no concurrency options at all:

```
Name:         random-cwvnj-1
Namespace:    default
Age:          25s
Image:        rhuss/random:1.0 (pinned to 946b7c)
Concurrency:
  Limit:      0
Service:      random

Conditions:
  OK TYPE                  AGE REASON
  ++ Ready                 21s
  ++ ContainerHealthy      21s
  ++ ResourcesAvailable    21s
  ++ Active                21s
```

File: pkg/kn/commands/revision/describe.go [hunk 5]
```diff
@@ -46,19 +62,225 @@ func NewRevisionDescribeCommand(p *commands.KnParams) *cobra.Command {
...
+		return
+	}
+	requests := c.Resources.Requests
+	limits := c.Resources.Limits
+	writeResourcesHelper(dw, "Memory", requests.Memory(), limits.Memory())
+	writeResourcesHelper(dw, "CPU", requests.Cpu(), limits.Cpu())
+}
+
+// Write request ... limits or only one of them
+func writeResourcesHelper(dw printers.PrefixWriter, label string, request *resource.Quantity, limit *resource.Quantity) {
```
Comment (rhuss) [inline pkg/kn/commands/revision/describe.go]: This method should be unified with the printing code for `service describe`.

File: pkg/kn/commands/revision/describe.go [hunk 6]
```diff
@@ -46,19 +62,225 @@ func NewRevisionDescribeCommand(p *commands.KnParams) *cobra.Command {
...
+
+	if maxScale != nil {
+		ret += strconv.Itoa(*maxScale)
+	} else {
+		ret += "∞"
+	}
+	return ret
+}
+
+func stringifyEnv(revision *v1alpha1.Revision) []string {
```
Comment (rhuss) [inline pkg/kn/commands/revision/describe.go]: unify with `service describe`.

File: pkg/kn/commands/revision/describe.go (hunk 1 above)
Comment (sixolet) [reply pkg/kn/commands/revision/describe.go]: sure

File: pkg/kn/commands/describe.go (hunk 2 above)
Comment (sixolet) [reply pkg/kn/commands/describe.go]: Done

File: pkg/kn/commands/revision/describe.go (hunk 3 above)
Comment (sixolet) [reply pkg/kn/commands/revision/describe.go]: Oh dang, it looks like I slightly misued git and forgot part of the refactor when I resurrected this PR. Resurrecting that :)

File: pkg/kn/commands/revision/describe.go [hunk 7]
```diff
@@ -46,19 +62,225 @@ func NewRevisionDescribeCommand(p *commands.KnParams) *cobra.Command {
...
+				if err != nil {
+					return err
+				}
+				return printer.PrintObj(revision, cmd.OutOrStdout())
 			}
-
-			err = printer.PrintObj(revision, cmd.OutOrStdout())
-			if err != nil {
-				return err
+			printDetails, err := cmd.Flags().GetBool("verbose")
```
Comment (maximilien) [inline pkg/kn/commands/revision/describe.go]: Why is this `err` not being checked?

File: pkg/kn/commands/revision/describe.go [hunk 8]
```diff
@@ -46,19 +62,225 @@ func NewRevisionDescribeCommand(p *commands.KnParams) *cobra.Command {
...
+func WriteEnv(dw printers.PrefixWriter, revision *v1alpha1.Revision, printDetails bool) {
+	env := stringifyEnv(revision)
+	if env != nil {
+		commands.WriteSliceDesc(dw, env, "Env", printDetails)
+	}
+}
+
+func WriteScale(dw printers.PrefixWriter, revision *v1alpha1.Revision) {
+	// Scale spec if given
+	scale, _ := clientserving.ScalingInfo(&revision.ObjectMeta)
```
Comment (maximilien) [inline pkg/kn/commands/revision/describe.go]: Don't want to nag but ignoring errors here... could easily be handled IMO.

**Pull request merged**